admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,534 @@
+                                  parameters_->ft_sensor_.frame_.external_, success);
+    ee_transform_ = get_transform(current_joint_state.positions, parameters_->kinematics_.tip_,
+                                  false, success);
+    world_transform_ = get_transform(current_joint_state.positions, parameters_->fixed_world_frame_.frame_.id_,
+                                     parameters_->fixed_world_frame_.frame_.external_, success);
+    sensor_transform_ = get_transform(current_joint_state.positions, parameters_->ft_sensor_.frame_.id_,
+                                      parameters_->ft_sensor_.frame_.external_, success);
+    cog_transform_ = get_transform(current_joint_state.positions, parameters_->gravity_compensation_.frame_.id_,
+                                   parameters_->gravity_compensation_.frame_.external_, success);
+
+    // get all needed rotations
+    ee_rot_ = ee_transform_.block<3, 3>(0, 0);
+    control_rot_ = control_transform_.block<3, 3>(0, 0);
+    sensor_rot_ = sensor_transform_.block<3, 3>(0, 0);
+    cog_rot_ = cog_transform_.block<3, 3>(0, 0);
+    world_rot_ = world_transform_.block<3, 3>(0, 0);
+
+    // apply filter and update wrench_world_ vector
+    process_wrench_measurements(measured_wrench, world_rot_ * sensor_rot_, world_rot_ * cog_rot_);
+
+    // transform wrench_world_ into base frame
+    Eigen::Matrix<double, 3, 2> wrench_base = world_rot_.transpose() * wrench_world_;
+
+    // calculate desired cartesian velocity in control frame
+    Eigen::Matrix<double, 3, 2> desired_vel = convert_joint_deltas_to_cartesian_deltas(current_joint_state.positions,
+                                                                                       current_joint_state.velocities,
+                                                                                       parameters_->ft_sensor_.frame_.id_,
+                                                                                       success);
+    // set desired_vel to zero if not using feed forward input
+    desired_vel = desired_vel * parameters_->use_feedforward_commanded_input_;
+
+    // Compute admittance control law
+    calculate_admittance_rule(wrench_base, desired_vel, dt);
+
+    // calculate drift due to integrating the joint positions
+    Eigen::Matrix<double, 3, 1> admittance_position_base = reference_ft_transform_.block<3,1>(0, 3) + admittance_position_.block<3,1>(0, 3);
+    Eigen::Matrix<double, 3, 1> admittance_actually_position_base = ft_transform_.block<3,1>(0,3);
+    Eigen::Matrix<double, 3, 3> ft_rot = ft_transform_.block<3,3>(0,0);
+    Eigen::Matrix<double, 3, 3> reference_ft_rot = reference_ft_transform_.block<3,3>(0,0);
+    Eigen::Matrix<double, 3, 3> admittance_rot = admittance_position_.block<3,3>(0,0);
+    Eigen::Matrix<double, 3, 3> R = admittance_rot.transpose()*(reference_ft_rot.transpose()*reference_ft_rot);
+
+    // calculate correction term
```

Comment:
The correction term is there because we want to support admittance control in position control mode. For a given admittance cartesian position x, there is no easy way to find the corresponding joint positions q. However, approximate the joint delta for every time step using the Jacobian and integrate them as time goes on. This approximation will drift over time, so a correction term is added to drive the drift to back to zero. There are other  ways this problem could be solved, but I found this solution to be the most suitable.